drivers/soundwire/intel.c

```diff
@@ -1412,6 +1412,108 @@ static int intel_master_remove(struct sdw_master_device *md)
 
 #ifdef CONFIG_PM
 
+static int _suspend(struct device *dev, bool clock_stop)
+{
+	struct sdw_cdns *cdns = dev_get_drvdata(dev);
+	struct sdw_intel *sdw = cdns_to_intel(cdns);
+	int ret;
+
+	if (cdns->bus.prop.hw_disabled) {
+		dev_dbg(dev, "SoundWire master %d is disabled, ignoring\n",
+			cdns->bus.link_id);
+		return 0;
+	}
+
+	dev_err(dev, "%s start\n", __func__);
+
+	if (clock_stop) {
+		ret = sdw_cdns_suspend(cdns, true);
+		if (ret < 0) {
+			dev_err(dev, "cannot enable clock stop on suspend\n");
+			return ret;
+		}
+	}
+
+	ret = sdw_cdns_enable_interrupt(cdns, false);
+	if (ret < 0) {
+		dev_err(dev, "cannot disable interrupts on suspend\n");
+		return ret;
+	}
+
+	ret = intel_link_power_down(sdw);
+	if (ret) {
+		dev_err(dev, "Link power down failed: %d", ret);
+		return ret;
+	}
+
+	/* set WAKEEN interrupt for wake-up events */
+	intel_shim_wake(sdw, clock_stop);
+
+	dev_err(dev, "%s done\n", __func__);
+
+	return 0;
+}
+
+static int _resume(struct device *dev, bool clock_stop)
+{
+	struct sdw_cdns *cdns = dev_get_drvdata(dev);
+	struct sdw_intel *sdw = cdns_to_intel(cdns);
+	int ret;
+
+	if (cdns->bus.prop.hw_disabled) {
+		dev_dbg(dev, "SoundWire master %d is disabled, ignoring\n",
+			cdns->bus.link_id);
+		return 0;
+	}
+
+	dev_err(dev, "%s start\n", __func__);
+
+	/* Invoke shim for wake disable */
+	intel_shim_wake(sdw, false);
+
+	ret = intel_init(sdw);
+	if (ret) {
+		dev_err(dev, "%s failed: %d", __func__, ret);
+		return ret;
+	}
+
+	/*
+	 * make sure all Slaves are tagged as UNATTACHED and provide
+	 * reason for reinitialization
+	 */
+	sdw_clear_slave_status(&sdw->cdns.bus,
+			       SDW_UNATTACH_REQUEST_MASTER_RESET);
+
+	ret = sdw_cdns_enable_interrupt(cdns, true);
+	if (ret < 0) {
+		dev_err(dev, "cannot enable interrupts during resume\n");
+		return ret;
+	}
+
+	/*
+	 * master is powered off on intel platforms, so it
+	 * should be reset for master initialization.
+	 */
+	ret = sdw_cdns_exit_reset(cdns);
```

Comment:
no. this is not needed when exiting clock stop mode.

Reply:
@plbossart   If this function is not invoked,  I got IO transfer timeout.  Master is powered off and it can't maintain its context. This is different with clock stop in SDW spec.  please check the following explanation in our HSD document:

Master IP being unpowered while Bus is under ClockStop is listed as scenario 9 in the table below. Reset exit from this case is detailed as scenarios 9 through 13 in the table below. Note that the sequence transitioning bus control as coordinated by SW is identical to those of scenarios 1 through 5, as the intent is to use the same flow for all reset exit cases: **initial power reset exit, hard reset exit (warm), reset exit after Clock Stop initiated by Slave device, or reset exit after Clock Stop triggered by SW.**

Reply:
same comment